test/expression_test.exs

```diff
@@ -413,4 +424,18 @@ defmodule ExpressionTest do
     assert "@@if(foo, bar, baz)" == Expression.escape("@if(foo, bar, baz)")
     assert "@@if(foo, bar.baz, baz)" == Expression.escape("@if(foo, bar.baz, baz)")
   end
+
+  test "parse!/1 skips expensive parsing is the expression is already an AST struct" do
+    # If the expression is a provided as a string then it gets parsed to an AST
+    assert [expression: [attribute: [atom: "contact", atom: "age"]], text: " + 2"] ==
+             Expression.parse!("@contact.age + 2")
+
+    # If the expression is a already provided as a parsed AST then it is simply
+    # returned as is
+    assert [expression: [attribute: [atom: "contact", atom: "age"]], text: " + 2"] ==
+             Expression.parse!(
```

Comment:
supplying AST (parsed structure) to a function called `parse!` seems a bit odd to be honest, what problem are we solving here?

Reply:
This is just because this function is called from inside every `evaluate_*` functions, so it was easier to add this check here then adding the check to every `evaluate_*` function, but I could also do that. I agree that it feels odd.

As per the PR description the goal here is to allow the library user to cache/store AST representation of commonnly evaluated expressions rather than forcing them to supply the expression as a string (and wasting time parsing it again) at every evaluation.